addons/mail/static/src/js/models/threads/mailbox.js

```diff
@@ -124,11 +124,11 @@ var Mailbox = SearchableThread.extend({
      */
     markAllMessagesAsRead: function (domain) {
         if (this._id === 'mailbox_inbox' && this.getMailboxCounter() > 0) {
+            this.trigger_up('move_messages_to_history', { messages: this.getMessages() });
```

Comment:
It should not add messages to history right away, after marking messages as read. Instead, it should wait the longpolling response that messages have been marked as read, similarly to how messages are removed from inbox.

Here is where messages are removed from Inbox from the `mark_as_read` notification on the longpolling:
https://github.com/odoo-dev/odoo/blob/master-mail-history-nla/addons/mail/static/src/js/services/mail_notification_manager.js#L327

The messages should be added to History here as well.